feat(gcoder): step length management when splitting curve

nojhan opened the pull request:
- Previous segment_to_gcode was splitting every curve in 1000, which may overstep the feed rate on some machines on small curves. This implementation starts by searching for a number of splits honoring (when possible) a configurable minimal step length.
- The step_length parameter is passed through the call stack and the --step-length argument is added to the interface.
- The CLI is put under __main__guard.
- WARNING -> NOTE for default parameters.
- Adds a --precision argument managing gcoder.precision.

SebKuzminsky commented:
Thank you for these features, they all look really useful and I want to merge them.

Unfortunately we both did some work in parallel, and now there's a merge conflict.  Please rebase your branch on top of the current master branch.  (I can help with this if you want.)

Also, your commit does many different things...  It's very much preferable if each commit does only one thing, so that they can be discussed and understood separately.  If possible, when you rebase, please split this big commit up into smaller logical commits.

Thanks!

nojhan committed:
feat(gcoder): add a minimal step_length

Previous segment_to_gcode was splitting every curve in 1000,
which may overstep the feed rate on some machines on small curves.
This implementation starts by searching for a number of splits honoring
(when possible) a configurable minimal step length.
Propagate the step_length parameter up the call stack.

nojhan committed:
fix(cli): log defaulting parameters as NOTE

nojhan committed:
feat(cli): adds --step-length and --precision

- Expose step_length from 0d404a4 as a CLI argument.
- Expose gcoder.precision as a CLI argument.

nojhan committed:
fix(cli): raise CLI code behind module guard

Allow importing the main file as a module without executing the CLI code.

nojhan commented:
This should be done with this forced push.

SebKuzminsky commented on gcoder.py:
```diff
@@ -305,8 +305,8 @@ def join_segments(this_seg, next_seg):
-                raise ValueError("segments are not even close to closed: %s, %s" % (this_seg, next_seg))
+            # else:
+                # raise ValueError("segments are not even close to closed: %s, %s" % (this_seg, next_seg))
         return (this_seg, next_seg)
```
Did you mean to comment this error check out?

SebKuzminsky commented on gcoder.py:
```diff
@@ -1046,21 +1046,37 @@ def path_segment_to_gcode(svg, segment, z=None):
+
+        start = svg.xy_to_mm(segment.start)
+        end = svg.xy_to_mm(segment.end)
+        stepdist = d(start,end)
```
I think this is the wrong distance metric to use.  Consider a nearly complete elliptic arc, or a bezier curve that nearly returns to its starting point - these would have a small `d(start, end)` distance (so this code would compute a small number of steps), but could have a long path distance (so the small number of steps would greatly over-simplify the path).

Instead, call the segment's `length()` method to get a better estimate of path length, then divide that by the desired step length to get the number of steps.

For example, consider this arc:
![out](https://github.com/SebKuzminsky/svg2gcode/assets/400548/543c5bb1-e583-4a25-a39f-c0d3ab144ec3)
```
$ python3 -c 'import svgpathtools; import math; a = svgpathtools.path.Arc((0+0j), (1+10j), 0, True, False, (0+1j)); print(math.sqrt((0-0)**2 + (0-1)**2)); print(a.length())'
1.0
39.63973762809824
```

SebKuzminsky commented on svg2gcode:
```diff
@@ -1076,6 +1084,7 @@ def load_job_file(jobfile):
     return jobs, jobfile_schema
 
 
+
```
Please avoid spurious whitespace changes.

SebKuzminsky commented on svg2gcode:
```diff
@@ -1162,14 +1176,12 @@ elif jobfile_schema == 2:
             for op in job['operations']:
                 results += handle_operation(op, path)
 
-
```
Please avoid spurious whitespace changes.

SebKuzminsky commented on svg2gcode:
```diff
@@ -1162,14 +1176,12 @@ elif jobfile_schema == 2:
     if type(result) == ResultPath:
         output_paths += result.path
 
-
```
Please avoid spurious whitespace changes.

SebKuzminsky commented on svg2gcode:
```diff
@@ -1121,6 +1130,11 @@ if args.z_approach == None:
 if args.z_top_of_material <= args.z_cut_depth:
     raise ValueError("--z-top-of-material (%f) is not above --z-cut-depth (%f)" % (args.z_top_of_material, args.z_cut_depth))
 
+gcoder.precision = args.precision
```
gcoder does not use precision anywhere, so this has no effect.

petterreinholdtsen commented:
@nojhan Do you plan to work on this pull request any more?

nojhan commented:
> @nojhan Do you plan to work on this pull request any more?

Probably not.
